builder/dockerfile/instructions/parse.go

```diff
@@ -0,0 +1,617 @@
+package instructions
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+	"github.com/docker/docker/builder/dockerfile/command"
+	"github.com/docker/docker/builder/dockerfile/parser"
+	"github.com/pkg/errors"
+)
+
+type parseRequest struct {
+	command    string
+	args       []string
+	attributes map[string]bool
+	flags      *BFlags
+	original   string
+}
+
+func nodeArgs(node *parser.Node) []string {
+	result := []string{}
+	for ; node.Next != nil; node = node.Next {
+		arg := node.Next
+		result = append(result, arg.Value)
+	}
+	return result
+}
+
+func newParseRequestFromNode(node *parser.Node) parseRequest {
+	return parseRequest{
+		command:    node.Value,
+		args:       nodeArgs(node),
+		attributes: node.Attributes,
+		original:   node.Original,
+		flags:      NewBFlagsWithArgs(node.Flags),
+	}
+}
+
+func ParseCommand(node *parser.Node) (interface{}, error) {
+	req := newParseRequestFromNode(node)
+	switch node.Value {
+	case command.Env:
+		return parseEnv(req)
+	case command.Maintainer:
+		return parseMaintainer(req)
+	case command.Label:
+		return parseLabel(req)
+	case command.Add:
+		return parseAdd(req)
+	case command.Copy:
+		return parseCopy(req)
+	case command.From:
+		return parseFrom(req)
+	case command.Onbuild:
+		return parseOnBuild(req)
+	case command.Workdir:
+		return parseWorkdir(req)
+	case command.Run:
+		return parseRun(req)
+	case command.Cmd:
+		return parseCmd(req)
+	case command.Healthcheck:
+		return parseHealthcheck(req)
+	case command.Entrypoint:
+		return parseEntrypoint(req)
+	case command.Expose:
+		return parseExpose(req)
+	case command.User:
+		return parseUser(req)
+	case command.Volume:
+		return parseVolume(req)
+	case command.StopSignal:
+		return parseStopSignal(req)
+	case command.Arg:
+		return parseArg(req)
+	case command.Shell:
+		return parseShell(req)
+	}
+
+	return nil, fmt.Errorf("unknown instruction: %s", strings.ToUpper(node.Value))
+}
+
+func Parse(ast *parser.Node) (stages []BuildableStage, metaArgs []ArgCommand, err error) {
+	for _, n := range ast.Children {
+		cmd, err := ParseCommand(n)
+		if err != nil {
+			return nil, nil, err
+		}
+		if len(stages) == 0 {
+			// meta arg case
+			if a, isArg := cmd.(*ArgCommand); isArg {
+				metaArgs = append(metaArgs, *a)
+				continue
+			}
+		}
+		switch c := cmd.(type) {
+		case *FromCommand:
+			if c.StageName == "" {
+				c.StageName = strconv.Itoa(len(stages))
+			}
+			stage := BuildableStage{Name: c.StageName, Commands: []interface{}{c}}
```

Comment:
Yes, I understand, my point is that a stageBuilder can be initialized by either a `FROM` command or a `ResumeBuild` command created when doing a `BuildFromConfig`. I like to leverage the typed-command style dispatching to change this alternative behavior.
Some other things to consider:
- From is not only about initializing the stage: At dispatch time, it might trigger `onbuild` expression dispatching
- From supports buildargs expansion and cannot be entirely resolved at parse time
- From can reference another stage, in this case it is not resolvable at parse time either, as it needs to get the runconfig produced by the referenced stage dispatching
- I'd like to keep the degree of consistency I have now: 1 dockerfile instruction == 1 typed command (even meta args are typed command, they just don't belong to a specific stage)
- In term of consistency, I really like the way it is, as the builder don't need to know anything about stage initialization, it only execute one list of commands per stage. The 1st command accidently initialize the stage runConfig, but the builder itself does not even care.